sonar-scanner-engine/src/main/java/org/sonar/scanner/sensor/DefaultSensorStorage.java

```diff
@@ -113,71 +118,148 @@ public DefaultSensorStorage(MetricFinder metricFinder, ModuleIssues moduleIssues
+        } else {
+          measureCache.put(component.key(), measure);
+        }
+      } else {
+        // Other coverage metrics are all integer values
+        DefaultMeasure<?> previousMeasure = measureCache.byMetric(component.key(), metric.key());
+        if (previousMeasure != null) {
+          measureCache.put(component.key(), new DefaultMeasure<Integer>()
+            .forMetric((Metric<Integer>) metric)
+            .withValue(Math.max((Integer) previousMeasure.value(), (Integer) measure.value())));
+        } else {
+          measureCache.put(component.key(), measure);
+        }
+      }
+    } else {
+      if (measureCache.contains(component.key(), metric.key())) {
+        throw new UnsupportedOperationException("Can not add the same measure twice on " + component + ": " + measure);
+      }
+      measureCache.put(component.key(), measure);
+    }
+  }
+
+  /**
+   * Merge the two line data measures, keeping max value in case they both contains a value for the same line.
+   */
+  private Map<Integer, Integer> mergeLineMetric(String value1, String value2) {
+    Map<Integer, Integer> data1 = KeyValueFormat.parseIntInt(value1);
+    Map<Integer, Integer> data2 = KeyValueFormat.parseIntInt(value2);
+    return Stream.of(data1, data2)
+      .map(Map::entrySet)
+      .flatMap(Collection::stream)
+      .collect(
+        Collectors.toMap(
+          Map.Entry::getKey,
+          Map.Entry::getValue,
+          Integer::max));
   }
 
-  private void setValueAccordingToMetricType(Measure<?> measure, org.sonar.api.measures.Metric<?> m, org.sonar.api.measures.Measure measureToSave) {
-    switch (m.getType()) {
-      case BOOL:
-        measureToSave.setValue(Boolean.TRUE.equals(measure.value()) ? 1.0 : 0.0);
-        break;
-      case INT:
-      case MILLISEC:
-      case WORK_DUR:
-      case FLOAT:
-      case PERCENT:
-      case RATING:
-        measureToSave.setValue(((Number) measure.value()).doubleValue());
-        break;
-      case STRING:
-      case LEVEL:
-      case DATA:
-      case DISTRIB:
-        measureToSave.setData((String) measure.value());
-        break;
-      default:
-        throw new UnsupportedOperationException("Unsupported type :" + m.getType());
+  public boolean isDeprecatedMetric(String metricKey) {
+    return DEPRECATED_METRICS_KEYS.contains(metricKey);
+  }
+
+  private boolean isLineMetrics(Metric<?> metric) {
+    return this.byLineMetrics.contains(metric);
+  }
+
+  public void validateCoverageMeasure(DefaultMeasure<?> measure, InputFile inputFile) {
+    Map<Integer, Integer> m = KeyValueFormat.parseIntInt((String) measure.value());
+    validatePositiveLine(m, inputFile.absolutePath());
+    validateMaxLine(m, inputFile);
+  }
+
+  private static void validateMaxLine(Map<Integer, Integer> m, InputFile inputFile) {
+    int maxLine = inputFile.lines();
+
+    for (int l : m.keySet()) {
```

Comment:
Of all the single-letter variable names, `l` is the worst, as it may look like the number 1 with some fonts, for example in GitHub diff view in Chrome...